앨범 도메인 리팩토링 진행

Hansolkkim opened the pull request:
close #50 

## 진행 내용
- [x] ColorPicker 관련 네이밍을 모두 ColorSelect로 변경 (다른 도메인들과의 통일성을 위해)
- [x] ColorSeletView의 CollectionViewCell의 "선택되어 있음"에 대한 로직 수정
- [x] 바인딩 로직이 너무 긴 경우 메소드로 분리
- [x] AlbumVM -> PhotoEditVM 변경 시에 모델 타입이 아닌 `BehaviorRelay<Album.Item>` 타입을 넘기도록 변경
- [x] Data -> UIImage 변환시 오버헤드 줄이도록 AlbumCell의 configure() 로직 변경

Hansolkkim committed:
rename: [#50] ColorPicker -> ColorSelect

- 다른 **PickerView 들과의 공통적인 명칭을 갖도록 ColorPicker와 관련된 모든 네이밍을 ColorSelect로 변경

Hansolkkim committed:
refactor: [#50] ColorSelectView의 선택되있는 셀 표시 로직 변경

- 기존에는 선택되어있는 Color 값을 viewModel.presetColors에서 찾아서 indexPath를 계산하여 UICollectionView.selectItem(_:_:_:)을 했는데 해당 로직이 너무 복잡하다고 판단하여, 셀을 그리는 때에 선택된 셀을 찾아서 표시해주는 로직을 추가

Hansolkkim committed:
refactor: [#50] 긴 바인딩 로직을 메소드로 분리

Hansolkkim committed:
chore: [#50] 기존의 바인딩 로직 분리 규칙에 맞게 코드 이동

Hansolkkim committed:
chore: [#50] ColorSelect 도메인에서 터치 반응 시점 변경

- UX 개선의 목적으로, 터치 반응 시점을 .recognized에서 .ended로 변경
- .recognized 시점에 터치 이벤트에 대한 반응이 나타나면 화면이 손에 가려 보이지 않는 상황에서 반응이 이루어져, 사용자 입장에서 어떤 반응이 발생하는지 확인하기 어렵다고 판단하여, 터치 시점을 위와 같이 변경

junu0516 commented on rabit/rabit.xcodeproj/project.pbxproj:
```diff
@@ -12,8 +12,8 @@
 		0232BD8E28ACC5DB009057F8 /* UIView + roundCorners.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0232BD8D28ACC5DB009057F8 /* UIView + roundCorners.swift */; };
-		0232BD9228AE1134009057F8 /* ColorPickerViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0232BD9128AE1134009057F8 /* ColorPickerViewController.swift */; };
-		0232BD9528AE11DD009057F8 /* ColorPickerViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0232BD9428AE11DD009057F8 /* ColorPickerViewModel.swift */; };
+		0232BD9228AE1134009057F8 /* ColorSelectViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0232BD9128AE1134009057F8 /* ColorSelectViewController.swift */; };
+		0232BD9528AE11DD009057F8 /* ColorSelectViewModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = 0232BD9428AE11DD009057F8 /* ColorSelectViewModel.swift */; };
```
이름 제가 바꿔도 되는데 ㅋㅋ.. 수고하셨어요!!

junu0516 commented on rabit/rabit/Album/ColorSelect/ColorSelectViewController.swift:
```diff
@@ -4,7 +4,7 @@ import RxSwift
 import RxGesture
 
-final class ColorPickerViewController: UIViewController {
+final class ColorSelectViewController: UIViewController {
```
요건 저도 preset -> selectView 같은 식으로 바꿔보겠습니다!

Hansolkkim committed:
refactor: [#50] AlbumVM-> PhotoEditVM 시에 참조타입 변수를 넘겨주도록 변경

- 기존에는 모델 타입을 넘겨주었기 때문에 변경된 Photo 데이터를 AlbumView에 적용하기 위해 Repository로부터 모든 데이터를 새로 받아왔으나, 참조타입을 넘겨줌으로써 변경된 Photo 데이터를 따로 받아오지 않고 변경된 Photo 데이터로부터 곧바로 변경시킬 수 있도록 개선.
- selectedPhoto의 변경을 중심으로 CollectionView를 reload시키기 떄문에, 불필요해진 AlbumNavigation.saveUpdatePhoto 변수 제거
- selectedPhoto의 변경을 중심으로 CollectionView를 reload시키기 위해, 터치된 셀의 IndexPath를 받아놓는 AlbumViewModel.indexSelected 변수 추가
- 기존에는 selectedPhoto의 변경이 직접적으로 navigation.showPhotoEditView와 바인딩되어 있었는데, PhotoEditView에서 photoStream을 통해 selectedPhoto를 변경시킬 수 있어져 불필요한 화면 전환이 발생함을 막기 위해, AlbumCollectionView의 Cell이 터치됐을 경우에만 화면 전환을 유도하기 위한 AlbumViewModel.showNextViewRequest 변수 추가

Hansolkkim committed:
refactor: [#50] AlbumCell의 Data -> UIImage 변환 로직 변경

- UIImage(data:)를 통해 Data를 UIImage로 변환하는 로직이 Main Thread에서 실행될 시에 오버헤드가 너무 커지기 떄문에, 이를 극복하고자 global queue에서 변환을 실행하고 UI 변경은 main thread에서 하도록하는 로직으로 변경

Hansolkkim committed:
chore: [#50] 불필요하게 제거한 코드 다시 추가

Hansolkkim committed:
fix: [#50] 터치 셀을 바꿀 경우에 collectionView가 reload되는 이슈 해결

- 첫 번째 셀을 누른 후 두 번째 셀을 누를 경우, 아무런 변경이 없음에도 collectionView가 reload되는 이슈 발견
- AlbumVM의 photoSelected 프로퍼티 변경에 따라 collectionView의 데이터소스가 변경되어 reload되는 흐름이었기 때문에 발생한 이슈임을 확인
- 이전에 삭제했던 "저장"버튼이 눌렸음을 알려주는 AlbumNavigation의 프로퍼티를 다시 살려, 해당 프로퍼티로 collectionView reload를 트리거하도록 수정
- 바인딩 로직의 정의 순서대로 실행됨을 확인하여, 맞는 순서대로 바인딩 로직 정의 순서를 변경 (또는 바인딩 메소드의 호출 순서 변경)

junu0516 commented on rabit/rabit/Album/AlbumViewController.swift:
```diff
@@ -128,9 +128,18 @@ private extension AlbumViewController {
             .bind(to: albumCollectionView.rx.items(dataSource: dataSource))
             .disposed(by: disposeBag)
 
+        albumCollectionView.rx.itemSelected
+            .bind(to: viewModel.indexSelected)
+            .disposed(by: disposeBag)
+
         albumCollectionView.rx.modelSelected(Album.Item.self)
             .bind(to: viewModel.photoSelected)
             .disposed(by: disposeBag)
+
+        albumCollectionView.rx.itemSelected
+            .map { _ in }
+            .bind(to: viewModel.showNextViewRequested)
+            .disposed(by: disposeBag)
```
취향차이라 반드시 수정할 필요는 없을듯한데
위에서 itemSelected에 대한 2개의 바인딩이
- 컬렉션뷰에서 선택된 셀의 인덱스값 정보를 뷰모델로 넘김
- 셀을 선택함에 따라 다음 뷰를 화면에 보여주기 위한 트리거 
의 2가지 동작을 동시에 처리하기 위함이면 do 연산과 bind 연산을 같이 사용해서 한 번의 바인딩으로 줄여볼 수도 있을 것 같아요!

Hansolkkim replied:
사실 두 바인딩이 아래처럼 하나로 합쳐져있었는데, `viewModel.indexSelected`의 변경과, `viewModel.photoSelected`의 변경과, 그리고 `viewModel.showNextViewRequested`의 변경이 순서적으로 의미가 있어 따로따로 작성한 부분이었습니다ㅎㅎ

먼저, **indexSelected**가 **photoSelected** 보다 먼저 변경되어야 "선택된 Cell을 변경"할 때에 최신의 indexPath를 사용할 수 있어지기 때문에 해당 순서가 의미가 있었습니다.
또한 **photoSelected**가 **navigaion.showNextViewRequested** 보다 먼저 변경되어야 PhotoEditView에게 최신의 Photo 데이터를 보내주어 화면에 그려줄 수 있기 때문에 해당 순서가 의미 있었습니다.

junu0516 commented on rabit/rabit/Album/Models/Photo.swift:
```diff
@@ -23,6 +23,16 @@ struct Photo: Equatable {
         self.color = color
     }
+
+    init() {
+        self.init(
+            categoryTitle: "",
+            goalTitle: "",
+            imageData: Data(),
+            date: Date(),
+            color: ""
+        )
+    }
```
이 부분은 디폴트 생성자로 처리하는 것도 괜찮아보입니다 ㅎㅎ

junu0516 commented on rabit/rabit/Album/PhotoEdit/PhotoEditViewModel.swift:
```diff
@@ -91,12 +90,17 @@ private extension PhotoEditViewModel {
-        selectedPhotoData
-            .map { $0 != selectedData }
-            .bind(to: saveButtonState)
+        saveButtonTouched.withLatestFrom(selectedPhotoData)
+            .bind(to: photoStream)
+            .disposed(by: disposeBag)
+
+        saveButtonTouched.withLatestFrom(selectedPhotoData)
+            .withUnretained(self)
+            .flatMapLatest { viewModel, data in
+                viewModel.albumRepository.updateAlbumData(data)
+            }
+            .bind(to: albumUpdateResult)
```
이 부분은 이렇게 이해하는게 맞는걸까요!?
- 저장버튼이 눌렸을 때, 전달받은 photoStream에 데이터를 넘겨서 화면에 수정된 이미지가 바로 적용되도록 함
- 동시에 Realm에 수정된 이미지 데이터를 저장시킴
(위의 2가지 동작을 동시에 처리하기 위한 의도인 지 확인차 여쭤봤습니다 ㅎㅎ)

Hansolkkim replied:
넵 맞습니다!

junu0516 commented on rabit/rabit/Album/AlbumViewController.swift:
```diff
@@ -128,14 +128,14 @@ private extension AlbumViewController {
-            .bind(to: viewModel.photoSelected)
-            .disposed(by: disposeBag)
-
         albumCollectionView.rx.itemSelected
             .bind(to: viewModel.indexSelected)
             .disposed(by: disposeBag)
 
+        albumCollectionView.rx.modelSelected(Album.Item.self)
+            .bind(to: viewModel.photoSelected)
+            .disposed(by: disposeBag)
+
```
이부분이 바인딩 순서 맞춰주셨단 부분이군요 ㅋㅋ
설명해주셔서 저도 처음 알았는데, 이렇게 코드 작성 순서에 따라서 바인딩 로직 실행 순서도 바뀌는게 신기하네요

Hansolkkim replied:
저도 이 부분 찾고 나니까 원하는대로 바인딩 로직 선언?호출?만 해주면 원하는 순서대로 값이 변경되어 조금 신기했어요ㅋㅋㅋ

junu0516 commented on rabit/rabit/Album/PhotoEdit/PhotoEditViewModel.swift:
```diff
@@ -91,16 +91,16 @@ private extension PhotoEditViewModel {
             .withUnretained(self)
             .flatMapLatest { viewModel, data in
                 viewModel.albumRepository.updateAlbumData(data)
             }
             .bind(to: albumUpdateResult)
             .disposed(by: disposeBag)
-
-        saveButtonTouched.withLatestFrom(selectedPhotoData)
-            .bind(to: photoStream)
-            .disposed(by: disposeBag)
```
이것도 혹시 순서 고려해서 바인딩 순서 고치신건가요!?
- 저장버튼 터치 후 Repository 계층 통해서 데이터 업데이트
- AlbumViewModel에서 전달받은 파라미터 업데이트
이런 순서로 제어한거 같아서 여쭤봤습니다 ㅎㅎ

Hansolkkim replied:
**albumUpdateResult**의 변경이 **navigation.didChangePhoto**를 트리거하고, 이것으로 인해 현재 **photoStream**의 값을 이용해 AlbumView의 AlbumCollectionView Cell이 업데이트되는 흐름입니다.

그런데, photoStream의 변경 이전에 albumUpdateResult의 변경이 되면, AlbumCollectionView Cell 업데이트 시에 보고있는 photoStream을 값이 "사용자에 의해 변경되기 전의 값"인 상태에서 업데이트를 하려고 하는 상황이 발생합니다.

따라서 photoStream을 먼저 변경하고 albumUpdateResult를 그 후에 변경하도록 하여, AlbumCollectionView Cell 업데이트 시에 최신의 photoStream 값을 사용하도록 하는걸 의도했습니다!

까먹고 있었는데 말씀주신 덕분에 찾아냈네요ㅋㅋㅋ

junu0516 reviewed (approved):
수고 많으셨어요 ㅋㅋㅋ 
노션 문서랑 같이 읽으면서 흐름 이해한게 맞나 여쭤보는 코멘트가 대부분입니다 😅
개인적으로 드는 생각은, 바인딩 코드 작성한 순서에 따라 런타임 시점에 이벤트 방출에 따른 로직 실행 순서도 바뀌는 점을 처음 알았는데, 이 부분을
```swift
do {
//우선 실행할 로직 정의
}
.bind {
//이벤트 방출 후 실행할 로직 정의
}
```
이런 흐름으로 순서를 좀 더 한눈에 알 수 있게 정의할 수 있는 부분은 없을까 하는 생각이 들어요!
사실 저도 화면 닫는 부분을 위와 같이 개선해볼 여지가 있다해서 학습하는중이라 ㅋㅋㅋ 저희 기능 합치고 한 번 기능 쭉 돌려보면서 이 부분 다시 언급해볼께요 ㅎㅎ
